src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt

```diff
@@ -57,26 +59,84 @@ class AuthService(
         return generateAccessToken(account)
     }
 
+    fun generateRecoveryToken(email: String): String? {
+        val account =
+            try {
+                accountService.getAccountByEmail(email)
+            } catch (e: Exception) {
+                logger.error(ErrorMessages.emailNotFound(email))
+                return null
+            }
+        return generateToken(
+            account,
+            Duration.ofMinutes(authConfigProperties.jwtRecoveryExpirationMinutes),
+            usePasswordHash = true
+        )
+    }
+
+    fun confirmRecoveryToken(recoveryToken: String, dto: PasswordRecoveryConfirmDto): Account? {
+        val jwt =
+            try {
+                jwtDecoder.decode(recoveryToken)
+            } catch (e: Exception) {
+                // throw InvalidBearerTokenException(ErrorMessages.invalidRecoveryToken)
+                logger.error(ErrorMessages.invalidRecoveryToken)
+                return null
+            }
+        if (jwt.expiresAt?.isBefore(Instant.now()) != false) {
+            throw InvalidBearerTokenException(ErrorMessages.expiredRecoveryToken)
+        }
+        val account = accountService.getAccountByEmail(jwt.subject)
+
+        val tokenPasswordHash =
+            try {
+                jwt.getClaim<String>("passwordHash")
+            } catch (e: Exception) {
+                // throw InvalidBearerTokenException(ErrorMessages.invalidRecoveryToken)
+                logger.error(ErrorMessages.invalidRecoveryToken)
+                return null
+            }
+
+        if (account.password != tokenPasswordHash) {
+            // throw InvalidBearerTokenException(ErrorMessages.invalidRecoveryToken)
+            logger.error(ErrorMessages.invalidRecoveryToken)
+            return null
+        }
+
+        account.password = passwordEncoder.encode(dto.password)
+        return accountService.updateAccount(account)
```

Comment:
@DoStini this is related to my previous comment but, if we want to hide invalid tokens then the user can distinguish between null and the account object. In that case, we should always return empty map on success

Reply:
But if we remove that, how can we tell if it is or isn't successful?